# Problem Set 3
Install packages phyloseq and microbiome. Complete the following problem set. Once completed, use File > Compile Report in R Studio to create HTML document of the problem set. Before you create the Report comment the install.package function.

#### 1: Install phyloseq and create a ps object with the attached files. Show the phyloseq-class experiment-level object

#### 2: List the following features about the ps object: 
sample names
min number of reads per sample
variable names
taxonomic ranks
summary of ps

sample_names(ps)
min(colSums(otu_table(ps)))
sample_variables(ps)
rank_names(ps)
summarize_phyloseq(ps)

#3: list all phyla in the ps 
get_taxa_unique(ps, "Phylum")

#4: Export table of ASVs with taxonomy
table_all<-cbind(tax_table(ps),otu_table(ps))
write.csv(table_all,file="ASV_abun.csv")

#5: subset samples BE and BT in factor region_c. Show ps
ps1 <- subset_samples(ps, region_c == "BE" | region_c ==  "BT")
ps1

#6: remove singletons. Show ps
ps1 <- prune_taxa(taxa_sums(ps1) > 1, ps1)
ps1

#7a: remove samples with less than 5000 reads. Show ps
ps1 = prune_samples(sample_sums(ps1) >= 5000, ps1)
ps1

#7b: remove ASVs whose mean value per sample is less than 1. Show ps
ps1 <- filter_taxa(ps1, function(x) mean(x) > 1, TRUE)
ps1

#8a: estimate absolute abundances and show results
head(otu_abs<- abundances(ps1),3)

#8b: estimate relative abundances and show results
head(otu_rel <- abundances(ps1, "compositional"),3)

#9a: Melt phyloseq data as a data frame table and show result
dfps <- psmelt(ps1)
head(dfps,4)

#9b: Show top 10 most abundant taxa
top_taxa(ps1, n = 10)

#10: transform phyloseq absolute abundances using the central log-ratio (CLR) transformation and show results
physeq_clr <- microbiome::transform(ps1, "clr")
head(otu_table(physeq_clr),3)
